alf/algorithms/vae.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+
+class VariationalAutoEncoder:
+    """
+    VariationalAutoEncoder encodes data into diagonal multivariate gaussian, do sampling with
```

Comment:
Because we are essential doing conditional VAE here, please comment the formulation of conditional VAE:  log p(x|y) = E_{z~q(z|x,y)} log P(x|z,y) - KL(q(z|x,y)||p(z|y)) and note that this class only compute z and  KL divergence.

Reply:
I don't think that it is conditional vae here,  though could be very easily turned into one. cvae does not have prior network, it uses class conditional/classes as both encoding features, and hidden dimension is also concat of conditional  and gaussian mean and var.

here we implemented two vae here.
1) vanilla vae.
2) vae with prior network.

Reply:
Anyway, we should clearly document the computation it performs using some equations.